src/init.jl

```diff
@@ -5,54 +5,94 @@ export context, context!, device!, device_reset!
 
 ## initialization
 
-# the default device new threads will use, set when switching devices.
-# this behavior differs from the CUDA Runtime, where device 0 is always used.
-const default_device = Ref{Union{Nothing,CuDevice}}(nothing)
-
-const thread_contexts = Union{Nothing,CuContext}[]
-
 # FIXME: support for flags (see `cudaSetDeviceFlags`)
 
 """
-    CUDAnative.initialize_context()
+    CUDAnative.prepare_cuda_call()
+
+Prepare state for calling CUDA API functions.
 
-Initialize a CUDA context for the current thread.
+Many CUDA APIs, like the CUDA driver API used by CUDAnative, use global thread-local state
+to determine, e.g., which device to use. With Julia however, code is grouped in tasks.
+Execution can switch between them, and tasks can be executing on (and in the future migrate
+between) different threads. To synchronize these two worlds, call this function before any
+CUDA API call to update thread-local state based on the current task.
 
-This function is designed to be very fast, and should be called befor almost every CUDA
-interaction.
+If you need to maintain your own thread-local state, subscribe to context and task switch
+events using [`CUDAnative.atcontextswitch`](@ref) and [`CUDAnative.attaskswitch`](@ref) for
+proper invalidation.
 """
-@inline function initialize_context()
+@inline function prepare_cuda_call()
     tid = Threads.threadid()
+    task = current_task()
+
+    # detect when a different task is now executing on a thread
+    if @inbounds thread_tasks[tid] != task
+        switched_tasks(tid, task)
+    end
+
+    # initialize a CUDA context when first executing on a thread
     if @inbounds thread_contexts[tid] === nothing
-        _initialize_context()
+        initialize_thread(tid)
     end
 
-    check_exceptions() # FIXME: This doesn't really belong here
+    check_exceptions()
 
     return
 end
 
-@noinline function _initialize_context()
-    @debug "Initializing CUDA on thread $(Threads.threadid())"
+# the default device new threads will use, set when switching devices.
+# this behavior differs from the CUDA Runtime, where device 0 is always used.
+const default_device = Ref{Union{Nothing,CuDevice}}(nothing)
+
+# CUDA uses thread-bound contexts, but calling CuCurrentContext all the time is expensive,
+# so we maintain our own thread-local state keeping track of the current context.
+const thread_contexts = Union{Nothing,CuContext}[]
+@noinline function initialize_thread(tid::Int)
     ctx = CuCurrentContext()
     dev = if ctx === nothing
         something(default_device[], CuDevice(0))
     else
+        # compatibility with externally-initialized contexts
         device()
     end
+
     device!(dev)
 end
 
+# Julia executes with tasks, so we need to keep track of the active task for each thread
+# in order to detect task switches and update the thread-local state accordingly.
+# doing so using task_local_storage is too expensive.
+const thread_tasks = Union{Nothing,Task}[]
```

Comment:
This will prevent GC from freeing no longer referenced tasks,

Reply:
Ah, right. But assuming the thread entry will get overwritten by a different task, this should be fine? Or we'd need to store `WeakRef`s here.

Reply:
Should mostly be fine. Still something that might look like a potential memory leak.

Reply:
Yes, I'm converting it into WeakRef.